Marginalize out discrete latent sites in HMC via enumeration

neerajprad opened the pull request:
Resolves #1128 via parallel enumeration.

 - If a model has discrete latents, they will be enumerated over in parallel. In such cases, we need to supply `max_iarange_nesting` as arg to HMC/NUTS constructor to reserve dims for parallel enumeration. Also in these cases, we do not allow for stray batch dims, i.e. batch dimensions need to be wrapped inside an `iarange`.
 - Tested on gmm and a latent bernoulli model. Also, added some cheap tests in `mcmc/test_valid_models.py` to assert error/ok behavior.

fritzo commented on pyro/infer/mcmc/hmc.py:
```diff
@@ -93,6 +108,35 @@ def _get_trace(self, z):
+
+        log_prob_sum = 0.
+        for log_prob in log_probs.values():
+            while log_prob.dim() > self.max_iarange_nesting:
```
nit: it would.be cheaper to reshape, something like:
```py
enum_dim = log_prob.dim() - self.max_iarange_nesting
if enum_dim > 0:
    log_prob = log_sum_exp(log_prob.reshape(-1, *log_prob.shape[enum_dim:]), dim=0)
```

neerajprad replied:
+1; will make the change.

fritzo commented:
Do I understand correctly that:
1. This PR implements parallel enumeration but not sequential enumeration.
2. The restriction to parallel enumeration is reasonable because HMC already makes the assumption that model structure is fixed, and the main use case of sequential enumeration is for models with dynamic structure.

fritzo commented on pyro/infer/mcmc/hmc.py:
```diff
@@ -93,6 +108,36 @@ def _get_trace(self, z):
+        for log_prob in log_probs.values():
+            enum_dim = log_prob.dim() - self.max_iarange_nesting
+            if enum_dim > 0:
+                log_prob = log_sum_exp(log_prob.reshape(-1, *log_prob.shape[enum_dim:]), dim=0)
```
I believe you'll need to do something fancier like `Dice.in_context()` here. At the moment, you're summing over only the local probabilities rather than all probabilities. This will fail for models with nested enumeration, e.g.
```py
@poutine.broadcast
def model(data):
    p = pyro.sample("p", dist.Uniform(0., 1.))
    y = pyro.sample("y", dist.Bernoulli(p))
    q = 0.5 + 0.25 * y
    with pyro.iarange("data", len(data)):
        z = pyro.sample("z", dist.Bernoulli(q))
        mean = 2 * z - 1
        pyro.sample("obs", dist.Normal(mean, 1.), obs=data)
```
In this case I believe you'll currently compute
```py
p(p) = p_unif(p)
     + sum(p_bern(y | p) for y in [0,1])
     + prod(sum(p_bern(zi | q(y)) for zi in [0,1] for y in [0,1])
            for i in range(len(data)))
```
which is not equal to the intended
```py
p(p) = p_unif(p)
     + sum(p_bern(y | p) * prod(sum(p_bern(zi | q(y)) for zi in [0,1])
                                for i in range(len(data)))
           for y in [0,1])
```
Could we add something like this as a unit test against a vetted `log_prob` computation?

neerajprad replied:
Thanks for this example, really helped me understand the issue with nested enumeration. I will add a test to validate the log prob computation.

fritzo commented:
Maybe it's worth moving `_compute_dice_elbo()` from traceenum_elbo.py into the `Dice` class as a `Dice.compute_elbo()` method, and then implementing what you want as a `Dice.compute_log_prob()` method? It seems like their code bodies are quite similar, but maybe their assumptions are sufficiently different for them to more naturally lie closer to their callers?

neerajprad commented:
> EDIT I've attempted to do some of this in #1251, hopefully that helps a bit.

Thanks for the PR, @fritzo! Let me go through this in more detail, and reply to your comments.

neerajprad commented:
> Do I understand correctly that:

> This PR implements parallel enumeration but not sequential enumeration. The restriction to parallel enumeration is reasonable because HMC already makes the assumption that model structure is fixed, and the main use case of sequential enumeration is for models with dynamic structure.

That's right. Sequential enumeration will greatly add to the already considerable computational overhead (since I think we'll need to run the integrator on each trace separately), making it not as useful practically.

eb8680 commented:
> Sequential enumeration will greatly add to the already considerable computational overhead (since I think we'll need to run the integrator on each trace separately), making it not as useful practically.

This could still be very useful for models whose log-probability computation graph after enumeration is static, since those can still be accelerated by the PyTorch JIT.  I don't think we need to handle the sequential case in this PR now.

neerajprad commented:
Updated the PR with the following:
 - The log prob computation used `Dice.in_context` to get the weights from the enumerated sites.
 - Added a test that validates the log prob computation for nested enumeration (can be computed by hand). Also modified the latent bernoulli test to catch the case explained by @fritzo.

I am seeing that using `Dice` and computing `score_parts` has increased the runtime by 40-50%, so there might be some opportunity for optimizing specifically for HMC.

neerajprad commented:
I could not use `Dice.compute_expectation` because the cost (probs of continuous sites) factors do not add up between different ordinals (but rather need to multiplied together). Similarly the dims within `max_iarange_nesting` need to be multiplied. If we use `log_probs` instead, that leads to something which isn't as general, and is basically what we have in `HMC._compute_trace_log_prob` currently.

fritzo commented on pyro/infer/mcmc/hmc.py:
```diff
@@ -93,6 +109,41 @@ def _get_trace(self, z):
+
+        log_probs = defaultdict(float)
+        for name, site in model_trace.nodes.items():
+            if site["type"] == "sample" and not site["fn"].has_enumerate_support:
```
I believe this incorrectly ignores discrete observe sites?

neerajprad replied:
Thanks for catching, will change and a test for this.

neerajprad commented on pyro/infer/mcmc/util.py:
```diff
@@ -0,0 +1,105 @@
+
+
+class EnumTraceProbEvaluator(object):
+    def __init__(self,
```
@frtizo - As discussed, I have removed the dependency on Dice, and moved the log prob computation to its own class to make it easy to test (test cases are added in `infer/mcmc/test_valid_models.py`). I think this should be self-explanatory (and very similar to the Dice logic), but let me know in case I may have overlooked some corner cases.

neerajprad replied:
Also, this should resolve the bugs in the previous implementation. I realized that the logic of only looking at the enumerated dims when computing the upstream log probs was wrong.

fritzo commented on pyro/infer/mcmc/util.py:
```diff
@@ -0,0 +1,105 @@
+                 model_trace,
+                 has_enumerable_sites=False,
+                 max_iarange_nesting=float("inf")):
+        """
```
nit: could you move this comment to a class-level comment rather than an `__init__` comment?

neerajprad replied:
My bad, will change.

fritzo commented on pyro/infer/mcmc/util.py:
```diff
@@ -0,0 +1,105 @@
+        self.has_enumerable_sites = has_enumerable_sites
+        self.max_iarange_nesting = max_iarange_nesting
+        self.log_probs = defaultdict(list)
+        self._log_factors_cache = {}
```
nit: It's cheaper and probably simpler to use a cache of pairs rather than a pair of caches.

fritzo commented on tests/infer/mcmc/test_valid_models.py:
```diff
@@ -0,0 +1,253 @@
+        with pyro.iarange("data", len(data)):
+            pyro.sample("obs", dist.Bernoulli(sigmoid(n)), obs=data)
+
+    conditioned_model = poutine.enum(
```
nit: why not add these to decorators above?
```py
@poutine.enum(first_available_dim=1)
@config_enumerate(default="parallel")
@poutine.condition(data={"p": ..., "n": ...})
@poutine.broadcast
def model(data):
    ...
```

fritzo commented on pyro/infer/mcmc/util.py:
```diff
@@ -0,0 +1,105 @@
+        # Sum up terms from predecessor, and gather leaf nodes.
+        leaves_log_probs = {}
+        for target_ordinal in sorted(self.log_probs.keys()):
+            leaves_log_probs[target_ordinal] = self.log_probs[target_ordinal]
```
nit: this is a bit confusing. Consider simplifying to
```py
for target_ordinal, log_prob in self.log_probs.items():
    if any(target_ordinal < other for other in self.log_probs):
        continue  # not a leaf
    predecessors, log_factors = self._get_predecessors_log_factors(target_ordinal)
    leaves_log_probs[target_ordinal] = sum(log_prob + log_factors)
```

fritzo commented on pyro/infer/mcmc/util.py:
```diff
@@ -0,0 +1,101 @@
+        if target_ordinal in self._predecessors_cache:
+            return self._predecessors_cache[target_ordinal]
+        log_factors = []
+        predecessors = set()
```
Am I correct that `predecessors` is no longer used, so we can remove this data structure?

neerajprad replied:
Fixed.

fritzo commented on pyro/infer/mcmc/util.py:
```diff
@@ -0,0 +1,101 @@
+        # - summing up the dims within `max_iarange_nesting`.
+        # (i.e. multiplying probs within independent batches).
+        log_prob_sum = 0.
+        for ordinal in leaves_log_probs:
```
nit: simplfy using `.items()`
```py
for ordinal, log_prob in leaves_log_probs.items():
    ...
```

fritzo commented on pyro/infer/mcmc/util.py:
```diff
@@ -0,0 +1,101 @@
+
+        # Sum up terms from predecessor, and gather leaf nodes.
+        leaves_log_probs = {}
+        for target_ordinal in sorted(self.log_probs.keys()):
```
nit: simplify using `.items()`
```py
for target_ordinal, log_prob in self.log_probs.items():
    if ...
        continue:
    log_factors = self._get_log_factors(target_ordinal)
    leaves_log_probs[target_ordinal] = log_prob + sum(log_factors)
```

fritzo commented:
> we just need to multiply the terms per independence context

I guess you could aggregate up the tree, starting with leaves and moving outward/upward. As you leave each independence context you `prod(sum(-))`? Another approach that I've taken in sequential enumeration is to explicitly scale for multiple counting by accumulating a `denominators` dict keyed on `ordinal`. Whatever works :smile:

neerajprad commented:
> I guess you could aggregate up the tree, starting with leaves and moving outward/upward. As you leave each independence context you prod(sum(-))?

This is the approach I have taken now. It cleanly solves a bunch of issues we were getting earlier, but this does not work for general DAGs (i.e., indep contexts with multiple parents).

neerajprad commented on pyro/infer/mcmc/util.py:
```diff
@@ -0,0 +1,137 @@
+from pyro.util import check_site_shape
+
+
+class EnumTraceProbEvaluator(object):
```
@fritzo - This is ready for another round of review. 😄The important change is confined to this class. In HMC we are calling the constructor once to populate the children nodes and compute which dims should be evaluated/enumerated out. When we call `log_prob` subsequently, we reuse this computed metadata to aggregate the log prob terms from the leaf nodes upwards.

fritzo commented on pyro/infer/mcmc/hmc.py:
```diff
@@ -93,6 +108,15 @@ def _get_trace(self, z):
 
+    @staticmethod
+    def _iter_latent_nodes(trace):
+        for name, node in sorted(trace.iter_stochastic_nodes(), key=lambda x: x[0]):
```
Why sort by name rather than the default order (by execution time)?

neerajprad replied:
Yeah, any fixed order is fine. The advantage to using names is that the dict based data structures that we are using can be sorted anytime by key names (otherwise, we'll need to store the key order somewhere and then sort using that order).

fritzo commented on pyro/infer/mcmc/util.py:
```diff
@@ -0,0 +1,137 @@
+                 max_iarange_nesting=float("inf")):
+        self.has_enumerable_sites = has_enumerable_sites
+        self.max_iarange_nesting = max_iarange_nesting
+        default_cond_stack = (CondIndepStackFrame(name="default", dim=0, size=0, counter=None),)
```
Why not the empty tuple?
```py
self.root = frozenset((()))
```

fritzo commented on pyro/infer/mcmc/util.py:
```diff
@@ -0,0 +1,137 @@
+        self.has_enumerable_sites = has_enumerable_sites
+        self.max_iarange_nesting = max_iarange_nesting
+        default_cond_stack = (CondIndepStackFrame(name="default", dim=0, size=0, counter=None),)
+        self.root = frozenset(default_cond_stack)
```
Why not the empty `frozenset()`?

neerajprad replied:
I think this can be removed now. I was initially doing something like `if frozenset() in frozenset([1])` (which evaluates to False), so this was to not have to put in special cases to deal with the root node. Will remove this now.

fritzo commented on pyro/infer/mcmc/util.py:
```diff
@@ -0,0 +1,137 @@
+        ordering = {}
+        for name, site in model_trace.nodes.items():
+            if site["type"] == "sample":
+                if len(site["cond_indep_stack"]) == 0:
```
If you make the `root` an empty frozenset then these four lines simplify to
```py
ordering[name] = frozenset(site["cond_indep_stack"])
```

neerajprad replied:
Changed, having a special default key isn't needed anymore.

fritzo replied:
or simply
```py
ordering = {name: frozenset(site["cond_indep_stack"])
            for name, site in model_trace.nodes.items()
            if site["type"] == "sample"}
```

fritzo commented on pyro/infer/mcmc/util.py:
```diff
@@ -0,0 +1,137 @@
+                    check_site_shape(site, self.max_iarange_nesting)
+                log_probs[ordering[name]].append(site["log_prob"])
+
+        for ordinal in log_probs:
```
nit: If you use `.items()` you can avoid half of the hashing of frozensets.
```py
for ordinal, terms in log_probs.items():
    self._log_probs[ordinal] = sum(terms)
```
Also, cool trick to avoid `0.0 + x` tensor ops!

fritzo commented on pyro/infer/mcmc/util.py:
```diff
@@ -0,0 +1,137 @@
+import torch
+from collections import defaultdict
+
+from pyro.distributions.util import log_sum_exp
```
Could we try to use the `logsumexp` form internally, since that's what PyTorch 0.4.1 calls it?

neerajprad replied:
Oh..sorry, I forgot to update. Will do so.

fritzo commented on pyro/infer/mcmc/util.py:
```diff
@@ -0,0 +1,137 @@
+        for ordinal in log_probs:
+            self._log_probs[ordinal] = sum(log_probs[ordinal])
+
+    def _reduce(self, ordinal, agg_log_prob=torch.tensor(0.)):
```
It's dangerous to create a `torch.tensor` at library load time, since it will not pick up `torch.set_default_tensor_type()` invocations. Instead I'd default to `None` and to an `if agg_log_prob is not None:` check.

neerajprad replied:
Good catch! Will change.

neerajprad replied:
I think it will be safe here to just use a python scalar. If a trace has no enumerable sites, we won't be going through this code path anyways, and if it has this 0. will get added as expected on the correct device.

fritzo commented on pyro/infer/mcmc/util.py:
```diff
@@ -0,0 +1,137 @@
+        """
+        Aggregate the `log_prob` terms using depth first search.
+        """
+        if self._children[ordinal] is None:
```
Hmm `self._children` is a `defaultdict(list)`. Do you instead mean to
```py
if not self._children[ordinal]:
    return self._reduce(ordinal)
```

fritzo commented on pyro/infer/mcmc/util.py:
```diff
@@ -0,0 +1,137 @@
+        """
+        if self._children[ordinal] is None:
+            return self._reduce(ordinal)
+        agg_log_prob = torch.tensor(0.)
```
You could avoid `torch.tensor` and save a tensor op by instead using python's `sum`:
```py
agg_log_prob = sum(map(self._aggregate_log_probs, self._children[ordinal]))
```

fritzo commented on pyro/infer/mcmc/util.py:
```diff
@@ -0,0 +1,137 @@
+        if not self.has_enumerable_sites:
+            return model_trace.log_prob_sum()
+        self._compute_log_prob_terms(model_trace)
+        return self._aggregate_log_probs(self.root).sum()
```
Do I understand correctly that:
1. all enumeration dims are to the left of `-max_iarange_nesting`, and they will have already been summed out by `self._aggregate_log_probs`
2. the result of `self._aggregate_log_probs(self.root)` will have non-reduced dims in exactly the global iaranges, i.e. the iaranges inside of which all sample statements are called, e.g. parallel particle iaranges?

fritzo commented on pyro/infer/mcmc/util.py:
```diff
@@ -0,0 +1,137 @@
+        enum_dims = set((i for i in range(-log_prob.dim(), -self.max_iarange_nesting)
+                         if log_prob.shape[i] > 1))
+        self._iarange_dims[ordinal] = iarange_dims
+        self._enum_dims[ordinal] = sorted(list(enum_dims - parent_enum_dims))
```
nit: no need for `list( )`

neerajprad replied:
done.

fritzo commented on pyro/infer/mcmc/util.py:
```diff
@@ -0,0 +1,137 @@
+        self._compute_log_prob_terms(model_trace)
+        # 1. Infer model structure - compute parent-child relationship.
+        sorted_ordinals = sorted(self._log_probs.keys())
+        for i in range(len(sorted_ordinals)):
```
nit: consider simplifying to
```py
for i, child_node in enumerate(sorted_ordinals):
```

neerajprad replied:
done.

neerajprad commented:
I think your comment got collapsed:
> Do I understand correctly that: all enumeration dims are to the left of -max_iarange_nesting, and they will have already been summed out by self._aggregate_log_probs

That's correct! 

> the result of self._aggregate_log_probs(self.root) will have non-reduced dims in exactly the global iaranges, i.e. the iaranges inside of which all sample statements are called, e.g. parallel particle iaranges?

They should be reduced dims, except that they will all be singleton, i.e. like `[[[1.3161]]]`. We don't currently have any global iaranges since we are not running parallel chains.

neerajprad commented:
> Can you confirm that you have tests that would have caught the double-counting errors that plagued earlier versions of this PR?

Yes, I have added many tests that would catch such errors in `test_valid_models`. There were a couple of tests that would have caught this even earlier, had I stand guilty of having double counted myself while writing the expected values. :) 

> After this merges, I'd like to add some more interesting models as tests, e.g. HMM and GMM.

There is a small gmm test in `test_hmc.py`, but it'll be great to have HMM and other interesting models.

neerajprad commented:
I think there is a dim error in the `logsumexp` implementation, which I'm looking into. Will update shortly.

neerajprad commented:
Thanks a ton for the detailed review, @fritzo. I think I have addressed all comments, and fixed the `logsumexp` bug now.

fritzo commented:
Nice work, @neerajprad!

martinjankowiak commented:
yaaay!
this is awesome @neerajprad !
also, seems like this PR also benefited from some serious CR from @fritzo !

eb8680 commented:
Awesome!
